edward/inferences/renyi_divergence.py

```diff
@@ -0,0 +1,176 @@
+            (see [@li2016renyi] - section 4.2)
+        backward_pass: str, optional.
+            Backward pass mode to be used.
+            Options: 'min', 'max', 'full'
+            (see [@li2016renyi] - section 4.2)
+    """
+    self.n_samples = n_samples
+    self.alpha = alpha
+    self.backward_pass = backward_pass
+
+    return super(RenyiDivergence, self).initialize(*args, **kwargs)
+
+  def build_loss_and_gradients(self, var_list):
+    """Build the Renyi ELBO function.
+
+    Its automatic differentiation is a stochastic gradient of
+
+    $ \mcalL_{R}^{\alpha}(q; x) =
+            \frac{1}{1-\alpha} \log \dsE_{q} \left[
+                \left( \frac{p(x, z)}{q(z)}\right)^{1-\alpha} \right].$
+
+    It uses:
+        + Monte Carlo approximation of the ELBO [@li2016renyi].
+        + Reparameterization gradients [@kingma2014auto].
+        + Stochastic approximation of the joint distribution [@li2016renyi].
+
+    #### Notes
+        + If the model is not reparameterizable, it returns a
+        NotImplementedError.
+        + See Renyi Divergence Variational Inference [@li2016renyi] for
+        more details.
+    """
+    if self.is_reparameterizable:
+      p_log_prob = [0.0] * self.n_samples
+      q_log_prob = [0.0] * self.n_samples
+      base_scope = tf.get_default_graph().unique_name("inference") + '/'
+      for s in range(self.n_samples):
+        # Form dictionary in order to replace conditioning on prior or
+        # observed variable with conditioning on a specific value.
+        scope = base_scope \
+            + tf.get_default_graph().unique_name("sample")
+        dict_swap = {}
+        for x, qx in six.iteritems(self.data):
+          if isinstance(x, RandomVariable):
+            if isinstance(qx, RandomVariable):
+              qx_copy = copy(qx, scope=scope)
+              dict_swap[x] = qx_copy.value()
+            else:
+              dict_swap[x] = qx
+
+        for z, qz in six.iteritems(self.latent_vars):
+          # Copy q(z) to obtain new set of posterior samples.
+          qz_copy = copy(qz, scope=scope)
+          dict_swap[z] = qz_copy.value()
+          q_log_prob[s] += tf.reduce_sum(
+              self.scale.get(z, 1.0) * qz_copy.log_prob(dict_swap[z]))
+
+        for z in six.iterkeys(self.latent_vars):
+          z_copy = copy(z, dict_swap, scope=scope)
+          p_log_prob[s] += tf.reduce_sum(
+              self.scale.get(z, 1.0) * z_copy.log_prob(dict_swap[z]))
+
+        for x in six.iterkeys(self.data):
+          if isinstance(x, RandomVariable):
+            x_copy = copy(x, dict_swap, scope=scope)
+            p_log_prob[s] += tf.reduce_sum(
+                self.scale.get(x, 1.0) * x_copy.log_prob(dict_swap[x]))
+
+      log_ratios = [p - q for p, q in zip(p_log_prob, q_log_prob)]
+
+      if self.backward_pass == 'max':
+        log_ratios = tf.stack(log_ratios)
+        log_ratios = tf.reduce_max(log_ratios, 0)
+        loss = tf.reduce_mean(log_ratios)
```

Comment:
If I understood the code correctly, `log_ratios` when first created is a list of `n_samples` elements, where each element is a log ratio calculation per sample from q. For the min / max modes, we take the min / max of these log ratios, which is a scalar.

Is `tf.reduce_mean` for the loss needed? You can also remove the `tf.stack` line in the min and max cases in the same way you didn't use it for the `self.alpha \approx 1` case.

Reply:
You're right. Thanks for spotting this.